README.md
<h1 align="center">
pygdbmi - Get Structured Output from GDB's Machine Interface
</h1>

<p align="center">

<a href="https://github.com/cs01/pygdbmi/actions">
<img src="https://github.com/cs01/pygdbmi/workflows/Tests/badge.svg?branch=master" alt="Test status" /></a>

<a href="https://badge.fury.io/py/pygdbmi">
<img src="https://badge.fury.io/py/pygdbmi.svg" alt="PyPI version"/></a>

</p>

**Documentation** [https://cs01.github.io/pygdbmi](https://cs01.github.io/pygdbmi)

**Source Code** [https://github.com/cs01/pygdbmi](https://github.com/cs01/pygdbmi)

---

Python (**py**) [**gdb**](https://www.gnu.org/software/gdb/) machine interface [(**mi**)](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI.html)

> GDB/MI is a line based machine oriented text interface to GDB and is activated by specifying using the --interpreter command line option (see Mode Options). It is specifically intended to support the development of systems which use the debugger as just one small component of a larger system.

## What's in the box?

1.  A function to parse gdb machine interface string output and return structured data types (Python dicts) that are JSON serializable. Useful for writing the backend to a gdb frontend. For example, [gdbgui](https://github.com/cs01/gdbgui) uses pygdbmi on the backend.
2.  A Python class to control and interact with gdb as a subprocess

To get [machine interface](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI.html) output from gdb, run gdb with the `--interpreter=mi2` flag like so:

```
gdb --interpreter=mi2
```

## Installation

    pip install pygdbmi

## Compatibility

### Operating Systems

Cross platform support for Linux, macOS and Windows

- Linux/Unix

  Ubuntu 14.04 and 16.04 have been tested to work. Other versions likely work as well.

- macOS

  Note: the error `please check gdb is codesigned - see taskgated(8)` can be fixed by codesigning gdb with [these instructions](http://andresabino.com/2015/04/14/codesign-gdb-on-mac-os-x-yosemite-10-10-2/). If the error is not fixed, please [create an issue in github](https://github.com/cs01/pygdbmi/issues).

- Windows

  Windows 10 has been tested to work with MinGW and cygwin.

### gdb versions

- gdb 7.6+ has been tested. Older versions may work as well.

## Examples

gdb mi defines a syntax for its output that is suitable for machine readability and scripting: [example output](https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Simple-Examples.html#GDB_002fMI-Simple-Examples):

```
-> -break-insert main
<- ^done,bkpt={number="1",type="breakpoint",disp="keep",
enabled="y",addr="0x08048564",func="main",file="myprog.c",
fullname="/home/myprog.c",line="68",thread-groups=["i1"],
times="0"}
<- (gdb)
```

Use `pygdbmi.gdbmiparser.parse_response` to turn that string output into a JSON serializable dictionary

```python
from pygdbmi import gdbmiparser
from pprint import pprint
response = gdbmiparser.parse_response('^done,bkpt={number="1",type="breakpoint",disp="keep", enabled="y",addr="0x08048564",func="main",file="myprog.c",fullname="/home/myprog.c",line="68",thread-groups=["i1"],times="0"')
pprint(response)
pprint(response)
# Prints:
# {'message': 'done',
#  'payload': {'bkpt': {'addr': '0x08048564',
#                       'disp': 'keep',
#                       'enabled': 'y',
#                       'file': 'myprog.c',
#                       'fullname': '/home/myprog.c',
#                       'func': 'main',
#                       'line': '68',
#                       'number': '1',
#                       'thread-groups': ['i1'],
#                       'times': '0',
#                       'type': 'breakpoint'}},
#  'token': None,
#  'type': 'result'}
```

## Programmatic Control Over gdb

But how do you get the gdb output into Python in the first place? If you want, `pygdbmi` also has a class to control gdb as subprocess. You can write commands, and get structured output back:

```python
from pygdbmi.gdbcontroller import GdbController
from pprint import pprint

# Start gdb process
gdbmi = GdbController()
print(gdbmi.command)  # print actual command run as subprocess
# Load binary a.out and get structured response
response = gdbmi.write('-file-exec-file a.out')
pprint(response)
# Prints:
# [{'message': 'thread-group-added',
#   'payload': {'id': 'i1'},
#   'stream': 'stdout',
#   'token': None,
#   'type': 'notify'},
#  {'message': 'done',
#   'payload': None,
#   'stream': 'stdout',
#   'token': None,
#   'type': 'result'}]
```

Now do whatever you want with gdb. All gdb commands, as well as gdb [machine interface commands](<(https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Input-Syntax.html#GDB_002fMI-Input-Syntax)>) are acceptable. gdb mi commands give better structured output that is machine readable, rather than gdb console output. mi commands begin with a `-`.

```python
response = gdbmi.write('-break-insert main')  # machine interface (MI) commands start with a '-'
response = gdbmi.write('break main')  # normal gdb commands work too, but the return value is slightly different
response = gdbmi.write('-exec-run')
response = gdbmi.write('run')
response = gdbmi.write('-exec-next', timeout_sec=0.1)  # the wait time can be modified from the default of 1 second
response = gdbmi.write('next')
response = gdbmi.write('next', raise_error_on_timeout=False)
response = gdbmi.write('next', raise_error_on_timeout=True, timeout_sec=0.01)
response = gdbmi.write('-exec-continue')
response = gdbmi.send_signal_to_gdb('SIGKILL')  # name of signal is okay
response = gdbmi.send_signal_to_gdb(2)  # value of signal is okay too
response = gdbmi.interrupt_gdb()  # sends SIGINT to gdb
response = gdbmi.write('continue')
response = gdbmi.exit()
```

## Parsed Output Format

Each parsed gdb response consists of a list of dictionaries. Each dictionary has keys `message`, `payload`, `token`, and `type`.

- `message` contains a textual message from gdb, which is not always present. When missing, this is `None`.
- `payload` contains the content of gdb's output, which can contain any of the following: `dictionary`, `list`, `string`. This too is not always present, and can be `None` depending on the response.
- `token` If an input command was prefixed with a (optional) token then the corresponding output for that command will also be prefixed by that same token. This field is only present for pygdbmi output types `nofity` and `result`. When missing, this is `None`.

The `type` is defined based on gdb's various [mi output record types](<(https://sourceware.org/gdb/onlinedocs/gdb/GDB_002fMI-Output-Records.html#GDB_002fMI-Output-Records)>), and can be

- `result` - the result of a gdb command, such as `done`, `running`, `error`, etc.
- `notify` - additional async changes that have occurred, such as breakpoint modified
- `console` - textual responses to cli commands
- `log` - debugging messages from gdb's internals
- `output` - output from target
- `target` - output from remote target
- `done` - when gdb has finished its output

## Contributing

Documentation fixes, bug fixes, performance improvements, and functional improvements are welcome. You may want to create an issue before beginning work to make sure I am interested in merging it to the master branch.

pygdbmi uses [nox](https://github.com/theacodes/nox) for automation.

See available tasks with

```
nox -l
```

Run tests and lint with

```
nox -s tests
nox -s lint
```

Positional arguments passed to `nox -s tests` are passed directly to `pytest`. For instance, to run only the parse tests use

```
nox -s tests -- tests/test_gdbmiparser.py
```

See [`pytest`'s documentation](https://docs.pytest.org/) for more details on how to run tests.

To format code using the correct settings use

```
nox -s format
```

Or, to format only specified files, use

```
nox -s format -- example.py pygdbmi/IoManager.py
```

### Making a release

Only maintainers of the [pygdbmi package on PyPi](https://pypi.org/project/pygdbmi/) can make a release.

In the following steps, replace these strings with the correct values:

- `<REMOTE>` is the name of the remote for the main pygdbmi repository (for instance, `origin`)
- `<VERSION>` is the version number chosen in step 2.

To make a release:

1. Checkout the `master` branch and pull from the main repository with `git pull <REMOTE> master`
2. Decide the version number for the new release; we follow
   [Semantic Versioning](https://semver.org/) but prefixing the version with `0.`: given a version
   number _0.SECOND.THIRD.FOURTH_, increment the:
   - _SECOND_ component when you make incompatible API changes
   - _THIRD_ component when you add functionality in a backwards compatible manner
   - _FOURTH_ component when you make backwards compatible bug fixes
3. Update `CHANGELOG.md` to list the chosen version number instead of `## dev`
4. Update `__version__` in `pygdbmi/__init__.py` to the chosen version number
5. Create a branch, for instance using `git checkout -b before-release-<VERSION>`
6. Commit your changes, for instance using `git commit -a -m 'Bump version to <VERSION> for release'`
7. Check that the docs look fine by serving them locally with `nox -s serve_docs`
8. Push the branch, for instance with `git push --set-upstream <REMOTE> before-release-<VERSION>`
9. If tests pass on the PR you created, you can merge into `master`
10. Go to the [new release page](https://github.com/cs01/pygdbmi/releases/new) and prepare the
    release:
    - Add a tag in the form `v<VERSION>` (for example `v0.1.2.3`)
    - Set the title to `pygdbmi v<VERSION>` (for example `pygdbmi v0.1.2.3`)
    - Copy and paste the section for the new release only from `CHANGELOG.md` excluding the line
      with the version number
    - Press “Publish release”
10. Publish the release to PyPI with `nox -s publish`
11. Publish the docs with `nox -s publish_docs`
11. Verify that the [PyPi page for pygdbmi](https://pypi.org/project/pygdbmi/) looks correct
12. Verify that the [published docs](https://cs01.github.io/pygdbmi/) look correct
13. Prepare for changes for the next release by adding something like this above the previous
    entries in `CHANGELOG.md` (where `<VERSION+1>` is `<VERSION>` with the last digit increaded
    by 1):

    ```
    ## <VERSION+1>.dev0

    - *Replace this line with new entries*
    ```

14. Create a branch for the changes with `git checkout -b after-release-<VERSION>`
15. Commit the change with `git commit -m 'Prepare for work on the next release' CHANGELOG.md`
16. Push the branch with `git push --set-upstream <REMOTE> after-release-<VERSION>`
17. If tests pass, merge into `master`

## Similar projects

- [tsgdbmi](https://github.com/Guyutongxue/tsgdbmi) A port of pygdbmi to TypeScript
- [danielzfranklin/gdbmi](https://github.com/danielzfranklin/gdbmi) A port of pygdbmi to Rust

## Projects Using pygdbmi

- [gdbgui](https://github.com/cs01/gdbgui) implements a browser-based frontend to gdb, using pygdbmi on the backend
- [PINCE](https://github.com/korcankaraokcu/PINCE) is a gdb frontend that aims to provide a reverse engineering tool and a reusable library focused on games. It uses pygdbmi to parse gdb/mi based output for some functions
- [avatar²](https://github.com/avatartwo/avatar2) is an orchestration framework for reversing and analysing firmware of embedded devices. It utilizes pygdbmi for internal communication to different analysis targets.
- [UDB](https://undo.io/udb) is a proprietary time-travel debugger for C and C++ based on GDB. It uses pygdbmi in its extensive test suite to parse the debugger's output.
- [pwndbg-gui](https://github.com/AlEscher/pwndbg-gui) is a user-friendly graphical interface for [pwndbg](https://github.com/pwndbg/pwndbg), a tool that simplifies exploit development and reverse engineering with GDB. It uses pygdbmi to interact with GDB and get structured responses.
- Know of another project? Create a PR and add it here.

## Authors

- [Chad Smith](https://github.com/cs01) (main author and creator).
- [Marco Barisione](http://www.barisione.org/) (co-maintainer).
- [The community](https://github.com/cs01/pygdbmi/graphs/contributors). Thanks especially to @mariusmue, @bobthekingofegypt, @mouuff, and @felipesere.


CHANGELOG.md
# pygdbmi release history

## <0.11.0.1>.dev0

- *Replace this line with new entries*
- Added a flag `return_on_result`, which causes an immediate return when a response of type 'result' is received,
instead of waiting until the timeout passes.

## 0.11.0.0

**Breaking changes**

- Removed `pygdbmi.IoManager.make_non_blocking` from the public API; it's unrelated and was not meant to be public

Other changes

- Fixed a bug where notifications without a payload were not recognized as such
- Invalid octal sequences produced by GDB are left unchanged instead of causing a `UnicodeDecodeError` (#64)
- Fix a crash on Windows by waiting for the GDB process to exit in `GdbController.exit`
- Added type annotations to the whole public API
- Updated the examples in `README.md` to use the current API and show the results printed by this version of pygdbmi (#69)

Internal changes

- Update and freeze dependencies for documentation generation
- Refactored the code to parse MI records to decrease the number of regex matches to perform
- Added `__all__` to all modules, which means that star imports (like `from pygdbmi.gdbmiparser import *`) will not pollute the namespace with modules used by pygdbmi itself
- Added `nox -s format` to re-format the source code using the correct options
- Reformatted all imports with `isort`, and use it as part of `nox -s lint` and `nox -s format`
- Converted tests to use pytest's test structure rather than the unittest-based one
- Added mypy configuration to detect more problems and to force all code to be annotated
- Added a test for `example.py`
- Replaced uses of `distutils.spawn.find_executable`, which is deprecated, with `shutil.which`
- Ran [`pyupgrade`](https://github.com/asottile/pyupgrade) (with option `--py37-plus`) on the codebase to convert to Python 3.7 idioms
- Excluded some common backup and cache files from `MANIFEST.in` to prevent unwanted files to be included which causes `check-manifest` to fail
- Fix `.flake8` to not cause errors with some versions of the `flake8` tool

## 0.10.0.2

- Strings containing escapes are now unescaped, both for messages in error records, which were previously mangled (#57), and textual records, which were previously left escaped (#58)
- Dropped support for Python 3.6 and added explicit support for Python 3.9 and 3.10.

## 0.10.0.1

- Fix bug with `time_to_check_for_additional_output_sec`, as it was not being used when passed to `GdbController`

## 0.10.0.0

**Breaking Changes**

- Drop support for Python 3.5
- Update `GdbController()` API. New API is `GdbController(command: Optional[List[str]], time_to_check_for_additional_output_sec: Optional[int])`.
- `GdbController.verify_valid_gdb_subprocess()` was removed
- Remove `NoGdbProcessError` error

Other Changes

- Add new `IoManager` class to handle more generic use-cases
- [dev] use pytest for testing
- gdb mi parsing remains unchanged

## 0.9.0.3

- Drop support for 2.7, 3.4
- Add support for 3.7, 3.8
- Add `py.typed` file so mypy can enforce type hints on `pygdbmi`
- Do not log in StringStream (#36)
- Updates to build and CI tests (use nox)
- Use mkdocs and mkdocstrings
- Doc updates

## 0.9.0.2

- More doc updates

## 0.9.0.1

- Update docs

## 0.9.0.0

- Stop buffering output
- Use logger in GdbController; modify `verbose` arguments.
- Remove support for Python 3.3

## 0.8.4.0

- Add method `get_subprocess_cmd` to view the gdb command run in the shell

## 0.8.3.0

- Improve reading gdb responses on unix (performance, bugfix) (@mouuff)

## 0.8.2.0

- Add support for [record and replay (rr) gdb supplement](http://rr-project.org/)

## 0.8.1.1

- Discard unexpected text from gdb

## 0.8.1.0

- Add native Windows support

## 0.8.0.0

- Make parsing more efficient when gdb outputs large strings
- Add new methods to GdbController class: `spawn_new_gdb_subprocess`, `send_signal_to_gdb`, and `interrupt_gdb`

## 0.7.4.5

- Update setup.py

## 0.7.4.4

- Fix windows ctypes import (#23, @rudolfwalter)

## 0.7.4.3

- Workaround gdb bug with repeated dictionary keys

## 0.7.4.2

- Improved buffering of incomplete gdb mi output (@trapito)
- Remove support of Python 3.2

## 0.7.4.1

- Preserve leading and trailing spaces in gdb/mi output (plus unit tests)
- Add unit test for buffering of gdb/mi output
- Documentation updates
- Refactoring

## 0.7.4.0

- Add more exception types (`NoGdbProcessError`, `GdbTimeoutError`)
- Add logic fixes for Windows (@johncf)
- Use codecs.open() to open the readme.rst, to prevent locale related bugs (@mariusmue)

## 0.7.3.3

- Add alternate pipe implementation for Windows

## 0.7.3.2

- Replace `epoll` with `select` for osx compatibility (@felipesere)

## 0.7.3.1

- Fix README

## 0.7.3.0

- Add support for gdb/mi (optional) tokens (@mariusmue)


noxfile.py
import itertools
import shutil
from pathlib import Path

import nox
from nox.sessions import Session


nox.options.sessions = ["tests", "lint", "docs"]
nox.options.reuse_existing_virtualenvs = True


# Run tests with (at least) the oldest and newest versions we support.
# If these are modified, also modify .github/workflows/tests.yml and the list of supported versions
# in setup.py.
@nox.session(python=["3.7", "3.10"])
def tests(session: Session) -> None:
    session.install(".", "pytest")
    session.run("pytest", *session.posargs)


LINTED_PATHS = {
    str(p.resolve())
    for p in itertools.chain(
        # All top-level Python files.
        Path(".").glob("*.py"),
        # Plus Python files in these specified directories.
        *(Path(d).glob("**/*.py") for d in ("pygdbmi", "tests"))
    )
}

ISORT_OPTIONS = ["--profile", "black", "--lines-after-imports", "2"]


# `format` is a builtin so the function is named differently.
@nox.session(name="format")
def format_(session: Session) -> None:
    """Re-format all Python source files or, if positionals are passed, only the
    specified files."""
    files = LINTED_PATHS
    if session.posargs:
        # Only use positional arguments which are linted files.
        files = files & {str(Path(f).resolve()) for f in session.posargs}

    session.install("isort", "black", "flake8", "mypy", "check-manifest")
    session.run("isort", *ISORT_OPTIONS, *files)
    session.run("black", *files)


@nox.session()
def lint(session: Session) -> None:
    session.install(
        # Packages needed as they are used directly.
        "black",
        "check-manifest",
        "flake8",
        "isort",
        "mypy",
        # Packages needed to provide types for mypy.
        "nox",
        "pytest",
    )
    session.run("isort", "--check-only", *ISORT_OPTIONS, *LINTED_PATHS)
    session.run("black", "--check", *LINTED_PATHS)
    session.run("flake8", *LINTED_PATHS)
    session.run("mypy", *LINTED_PATHS)
    session.run("check-manifest")
    session.run("python", "setup.py", "check", "--metadata", "--strict")


def install_mkdoc_dependencies(session: Session) -> None:
    session.install("-r", "mkdoc_requirements.txt")


@nox.session
def docs(session: Session) -> None:
    install_mkdoc_dependencies(session)
    session.run("mkdocs", "build")


@nox.session
def serve_docs(session: Session) -> None:
    install_mkdoc_dependencies(session)
    session.run("mkdocs", "serve")


@nox.session
def publish_docs(session: Session) -> None:
    install_mkdoc_dependencies(session)
    session.run("mkdocs", "gh-deploy")


@nox.session(python="3.7")
def build(session: Session) -> None:
    session.install("setuptools", "wheel", "twine")
    shutil.rmtree("dist", ignore_errors=True)
    shutil.rmtree("build", ignore_errors=True)
    session.run("python", "setup.py", "--quiet", "sdist", "bdist_wheel")
    session.run("twine", "check", "dist/*")


@nox.session(python="3.7")
def publish(session: Session) -> None:
    build(session)
    print("REMINDER: Has the changelog been updated?")
    session.run("python", "-m", "twine", "upload", "dist/*")
    publish_docs(session)
